File: src/CSharp/src/Durable7.FingerTree/AncestralSliceQueue.cs
using System.Collections;
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;

namespace Durable7.FingerTree;

/// <summary>
/// Represents an immutable, fully persistent queue slice as an interval on an append ancestry path.
/// </summary>
/// <typeparam name="T">The element type.</typeparam>
/// <remarks>
/// <para>
/// A value stores one arena, a tail node, the absolute depth of its first visible element, and a
/// redundant count cache.
/// Empty values retain the node immediately before their window as an anchor, so appending to any
/// empty slice yields exactly the new value without exposing a discarded prefix. All handles are
/// immutable; adding below an old handle creates a sibling branch in the shared monotone arena.
/// </para>
/// <para>
/// Let U be the arena's leaf-add time and Q its level-ancestor query time. <see cref="AddLast"/>
/// costs U; <see cref="First"/>, the indexer, <see cref="Take"/>, <see cref="Slice"/>, and
/// nontrivial <see cref="SplitAt"/> cost Q; <see cref="Last"/>, <see cref="RemoveFirst"/>,
/// <see cref="RemoveLast"/>, and <see cref="Drop"/> are O(1). Alstrup--Holm incremental ancestry
/// gives U = Q = O(1) worst case with linear arena space. <see cref="CreateMyers"/> instead uses
/// O(1)-amortized addition and O(log M) ancestor queries after M historical appends.
/// </para>
/// <para>
/// The operation set intentionally excludes prepend, point replacement, arbitrary middle edits,
/// and concatenation of unrelated histories. Arena space is charged to all successful historical
/// appends, not only to the visible length of one handle.
/// </para>
/// </remarks>
[DebuggerDisplay("Count = {Count}, LowDepth = {_lowDepth}, Tail = {_tail}")]
public sealed class AncestralSliceQueue<T> : IReadOnlyList<T>
{
    private readonly IIncrementalAncestorArena<T> _arena;
    private readonly int _tail;
    private readonly int _lowDepth;

    private AncestralSliceQueue(
        IIncrementalAncestorArena<T> arena,
        int tail,
        int lowDepth,
        int count)
    {
        _arena = arena;
        _tail = tail;
        _lowDepth = lowDepth;
        Count = count;
    }

    /// <summary>Gets the number of visible values.</summary>
    public int Count { get; }

    /// <summary>Gets whether this handle denotes an empty ancestry interval.</summary>
    public bool IsEmpty => Count == 0;

    /// <summary>Gets the first visible value.</summary>
    /// <exception cref="InvalidOperationException">The queue is empty.</exception>
    public T First
    {
        get
        {
            EnsureNotEmpty();
            var node = Count == 1 ? _tail : _arena.AncestorAtDepth(_tail, _lowDepth);
            return _arena.GetValue(node);
        }
    }

    /// <summary>Gets the last visible value.</summary>
    /// <exception cref="InvalidOperationException">The queue is empty.</exception>
    public T Last
    {
        get
        {
            EnsureNotEmpty();
            return _arena.GetValue(_tail);
        }
    }

    /// <summary>Gets the visible value at a zero-based index.</summary>
    /// <param name="index">The zero-based visible index.</param>
    /// <exception cref="ArgumentOutOfRangeException"><paramref name="index"/> is outside the queue.</exception>
    public T this[int index]
    {
        get
        {
            if ((uint)index >= (uint)Count)
                throw new ArgumentOutOfRangeException(nameof(index));
            var node = index == Count - 1
                ? _tail
                : _arena.AncestorAtDepth(_tail, checked(_lowDepth + index));
            return _arena.GetValue(node);
        }
    }

    /// <summary>Creates an empty queue owned by an explicit incremental-ancestor arena.</summary>
    /// <param name="arena">The manager that will retain and navigate every appended node.</param>
    /// <returns>An empty anchored queue.</returns>
    /// <exception cref="ArgumentNullException"><paramref name="arena"/> is null.</exception>
    /// <exception cref="ArgumentException">The arena's bottom node is not at depth -1.</exception>
    public static AncestralSliceQueue<T> Create(IIncrementalAncestorArena<T> arena)
    {
        ArgumentNullException.ThrowIfNull(arena);
        var bottom = arena.Bottom;
        if (arena.GetDepth(bottom) != -1)
            throw new ArgumentException("The arena bottom node must have depth -1.", nameof(arena));
        return new AncestralSliceQueue<T>(arena, bottom, lowDepth: 0, count: 0);
    }

    /// <summary>Creates an empty queue backed by the shipped Myers jump-link arena.</summary>
    /// <returns>An empty queue with a fresh, independently collectible arena.</returns>
    public static AncestralSliceQueue<T> CreateMyers() => Create(new MyersIncrementalAncestorArena<T>());

    /// <summary>Creates a Myers-backed queue from values in enumeration order.</summary>
    /// <param name="values">The values to append.</param>
    /// <returns>A queue containing the values.</returns>
    public static AncestralSliceQueue<T> CreateRange(IEnumerable<T> values)
    {
        ArgumentNullException.ThrowIfNull(values);
        var result = CreateMyers();
        foreach (var value in values)
            result = result.AddLast(value);
        return result;
    }

    /// <summary>Returns a queue with one value appended to this handle's current tail or empty anchor.</summary>
    /// <param name="value">The value to append.</param>
    /// <returns>The extended persistent branch.</returns>
    /// <exception cref="OverflowException">The visible count or ancestry depth cannot grow further.</exception>
    public AncestralSliceQueue<T> AddLast(T value)
    {
        if (Count == int.MaxValue)
            throw new OverflowException("The queue count exceeds Int32.MaxValue.");
        var tail = _arena.AddLeaf(_tail, value);
        return new AncestralSliceQueue<T>(_arena, tail, _lowDepth, Count + 1);
    }

    /// <summary>Returns the queue without its first visible value.</summary>
    /// <returns>The remaining ancestry interval.</returns>
    /// <exception cref="InvalidOperationException">The queue is empty.</exception>
    public AncestralSliceQueue<T> RemoveFirst()
    {
        EnsureNotEmpty();
        return new AncestralSliceQueue<T>(_arena, _tail, checked(_lowDepth + 1), Count - 1);
    }

    /// <summary>Returns the queue without its last visible value.</summary>
    /// <returns>The remaining ancestry interval.</returns>
    /// <exception cref="InvalidOperationException">The queue is empty.</exception>
    public AncestralSliceQueue<T> RemoveLast()
    {
        EnsureNotEmpty();
        return new AncestralSliceQueue<T>(_arena, _arena.GetParent(_tail), _lowDepth, Count - 1);
    }

    /// <summary>Attempts to remove and return the first visible value.</summary>
    /// <param name="value">The first value on success.</param>
    /// <param name="result">The remaining queue on success; otherwise, this queue.</param>
    /// <returns><see langword="true"/> when a value was removed.</returns>
    public bool TryRemoveFirst(
        [MaybeNullWhen(false)] out T value,
        out AncestralSliceQueue<T> result)
    {
        if (IsEmpty)
        {
            value = default;
            result = this;
            return false;
        }

        value = First;
        result = RemoveFirst();
        return true;
    }

    /// <summary>Attempts to remove and return the last visible value.</summary>
    /// <param name="value">The last value on success.</param>
    /// <param name="result">The remaining queue on success; otherwise, this queue.</param>
    /// <returns><see langword="true"/> when a value was removed.</returns>
    public bool TryRemoveLast(
        [MaybeNullWhen(false)] out T value,
        out AncestralSliceQueue<T> result)
    {
        if (IsEmpty)
        {
            value = default;
            result = this;
            return false;
        }

        value = Last;
        result = RemoveLast();
        return true;
    }

    /// <summary>Returns the first <paramref name="count"/> visible values.</summary>
    /// <param name="count">The prefix length.</param>
    /// <returns>An appendable ancestry prefix.</returns>
    public AncestralSliceQueue<T> Take(int count)
    {
        ValidateBoundary(count, nameof(count));
        return count == Count ? this : Slice(0, count);
    }

    /// <summary>Returns the queue after omitting its first <paramref name="count"/> visible values.</summary>
    /// <param name="count">The number of values to omit.</param>
    /// <returns>An appendable ancestry suffix.</returns>
    public AncestralSliceQueue<T> Drop(int count)
    {
        ValidateBoundary(count, nameof(count));
        if (count == 0)
            return this;
        return new AncestralSliceQueue<T>(
            _arena,
            _tail,
            checked(_lowDepth + count),
            Count - count);
    }

    /// <summary>Returns one contiguous, appendable ancestry slice.</summary>
    /// <param name="index">The zero-based first visible index.</param>
    /// <param name="count">The number of visible values in the result.</param>
    /// <returns>The selected ancestry interval.</returns>
    public AncestralSliceQueue<T> Slice(int index, int count)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(index);
        ArgumentOutOfRangeException.ThrowIfNegative(count);
        if (index > Count)
            throw new ArgumentOutOfRangeException(nameof(index));
        if (index > Count - count)
            throw new ArgumentOutOfRangeException(nameof(count));
        if (index == 0 && count == Count)
            return this;

        var lowDepth = checked(_lowDepth + index);
        var targetDepth = checked(lowDepth + count - 1);
        var currentTailDepth = checked(_lowDepth + Count - 1);
        var tail = targetDepth == currentTailDepth
            ? _tail
            : _arena.AncestorAtDepth(_tail, targetDepth);
        return new AncestralSliceQueue<T>(_arena, tail, lowDepth, count);
    }

    /// <summary>Splits this queue at a zero-based boundary.</summary>
    /// <param name="index">The number of visible values placed in the left result.</param>
    /// <returns>The appendable prefix and suffix.</returns>
    public (AncestralSliceQueue<T> Left, AncestralSliceQueue<T> Right) SplitAt(int index)
    {
        ValidateBoundary(index, nameof(index));
        return (Take(index), Drop(index));
    }

    /// <summary>Returns a stable front-to-back enumerator over this immutable ancestry interval.</summary>
    /// <returns>An enumerator.</returns>
    /// <remarks>
    /// The shipped implementation follows parent links into a temporary array, taking Theta(Count)
    /// time and Theta(Count) transient element slots. Concurrent appends cannot change the captured path.
    /// </remarks>
    public IEnumerator<T> GetEnumerator() => Enumerate().GetEnumerator();

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    private IEnumerable<T> Enumerate()
    {
        if (Count == 0)
            yield break;

        var values = new T[Count];
        var node = _tail;
        for (var index = Count - 1; index >= 0; index--)
        {
            values[index] = _arena.GetValue(node);
            if (index != 0)
                node = _arena.GetParent(node);
        }

        foreach (var value in values)
            yield return value;
    }

    private void ValidateBoundary(int index, string parameterName)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(index, parameterName);
        if (index > Count)
            throw new ArgumentOutOfRangeException(parameterName);
    }

    private void EnsureNotEmpty()
    {
        if (IsEmpty)
            throw new InvalidOperationException("The ancestral slice queue is empty.");
    }
}

File: src/Kotlin/FingerTree/src/durable7/fingertree/AncestralSliceQueue.kt
/*
 * A persistent queue whose every version is one contiguous interval of an append-only ancestry path.
 *
 * A value owns no nodes. It is the constant-sized pair (tail, lowDepth) naming an interval of the
 * unique bottom-to-tail path inside a shared incremental-ancestor arena, plus a redundant count
 * cache, so the visible values are the labels of the ancestors of the tail at absolute depths
 * lowDepth through depthOf(tail). Appending publishes one child below the tail; removing and slicing
 * move only the two interval boundaries. Because the arena is append-only and its published nodes
 * are immutable, every retained handle keeps denoting the same sequence forever, and adding below an
 * old handle starts a sibling branch instead of disturbing it.
 */
package durable7.fingertree

/**
 * One removed endpoint value together with the persistent queue that remains without it.
 *
 * The wrapper keeps a successful removal distinguishable from an empty queue even when the element
 * type is nullable, following this workspace's [RrbPop] precedent.
 *
 * @param T the element type.
 * @property value the removed endpoint value.
 * @property rest the persistent queue remaining after that endpoint is removed.
 */
public data class AncestralSlicePop<T>(
    public val value: T,
    public val rest: AncestralSliceQueue<T>,
)

/**
 * The two appendable ancestry intervals produced by [AncestralSliceQueue.splitAt].
 *
 * @param T the element type.
 * @property left the prefix holding the first `index` visible values.
 * @property right the suffix holding the remaining visible values.
 */
public data class AncestralSliceSplit<T>(
    public val left: AncestralSliceQueue<T>,
    public val right: AncestralSliceQueue<T>,
)

/**
 * Independently recomputed ancestry-interval metadata.
 *
 * @property count the cached visible count the validated handle reported.
 * @property lowDepth the absolute arena depth of the first visible value, which for an empty window
 *   is one greater than the depth of its retained anchor.
 * @property tailDepth the absolute arena depth of the retained tail node; it always equals
 *   `lowDepth + count - 1`, which for an empty window is the anchored-empty rule as an equation.
 * @property walkedNodeCount the number of ancestry nodes reached by following parent links from the
 *   tail, recomputed rather than read from the cached count.
 * @property publishedNodeCount the arena's labelled-node count when the validation ran.
 */
public data class AncestralSliceQueueStatistics(
    public val count: Int,
    public val lowDepth: Int,
    public val tailDepth: Int,
    public val walkedNodeCount: Int,
    public val publishedNodeCount: Int,
)

/**
 * An immutable, fully persistent queue slice denoting one interval of an append ancestry path.
 *
 * A value retains its arena, a tail node, the absolute depth of its first visible element, and a
 * redundant count cache. Empty values retain the node immediately before their window as an anchor,
 * so appending to any empty slice yields exactly the new value without exposing a discarded prefix;
 * a queue drained from the front and a queue drained from the back are both empty yet keep different
 * anchors. All handles are immutable, and adding below an old handle creates a sibling branch in the
 * shared monotone arena. Two handles may enumerate equal values while retaining different tails,
 * anchors, or arenas, so this type promises no constant-time sequence equality, no hashing, and no
 * canonical empty identity, and implements none of them.
 *
 * The bounds are parameterized by the backend. Let `M` be the number of labelled nodes ever
 * published by the arena, `U(M)` the cost of adding a leaf, and `Q(M)` the cost of one level-ancestor
 * query. Then [addLast] costs `U(M)`; [first], [get], [take], [slice], [popFirst], and a nontrivial
 * [splitAt] cost `Q(M)`; and [size], [isEmpty], [last], [removeFirst], [removeLast], [popLast], and
 * [drop] are O(1). Boundary-specialized calls make no ancestor query at all: `take(size)`, `drop(0)`,
 * the whole-range slice, any slice whose result ends at the retained tail, an index at the last
 * visible position, a singleton's [first], and a split at [size]. A split at `0` of a non-empty queue
 * is not query-free, because the anchored-empty rule requires the empty prefix to retain the node at
 * depth `lowDepth - 1` and only an ancestor query can name that node from the retained tail; on an
 * empty queue both split boundaries coincide with the whole range, so that split is query-free.
 *
 * Instantiating the backend with Alstrup-Holm incremental level ancestry would give
 * `U(M) = Q(M) = O(1)` worst case in linear arena space. That backend is not implemented here; the
 * statement is a reduction to a published result. The shipped [MyersIncrementalAncestorArena] has
 * O(1)-amortized leaf addition and O(log M) worst-case ancestor queries, so no amortized bound above
 * may be read as a worst-case bound.
 *
 * The algebra intentionally excludes prepending, point replacement, arbitrary middle edits, and
 * concatenation of unrelated histories; omitting them is precisely why the indexed and slicing bounds
 * are possible. Arena space is charged to all successful historical appends, not to the visible
 * length of one handle: dropping a prefix or releasing a handle reclaims nothing, and every published
 * value stays reachable until the arena itself becomes collectible.
 *
 * Handles are immutable and are safe for concurrent readers to the same extent as their arena and
 * their caller-owned elements. The shipped arena serializes its reads and writes on one private
 * monitor; that is a semantic thread-safety property, not a parallel-progress guarantee.
 *
 * @param T the element type.
 */
public class AncestralSliceQueue<T> private constructor(
    private val arena: IncrementalAncestorArena<T>,
    private val tail: Int,
    private val lowDepth: Int,
    count: Int,
) : Iterable<T> {
    /** Factories for anchored queues over a shipped or caller-supplied ancestor arena. */
    public companion object {
        /**
         * Creates an empty queue owned by an explicit incremental-ancestor arena.
         *
         * This is the public extension seam: supplying a backend with different `U(M)`/`Q(M)` costs
         * changes every parameterized bound this type documents. The arena retains and navigates
         * every appended node. This factory is O(1) and publishes nothing.
         *
         * @param arena the manager that will retain and navigate every appended node.
         * @throws IllegalArgumentException when the arena's bottom node is not at depth `-1`, so
         *   that no anchored-empty queue can be expressed over it.
         */
        public fun <T> empty(arena: IncrementalAncestorArena<T>): AncestralSliceQueue<T> {
            val bottom = arena.bottom
            require(arena.depthOf(bottom) == -1) { "The arena bottom node must have depth -1." }
            return AncestralSliceQueue(arena, bottom, lowDepth = 0, count = 0)
        }

        /**
         * Creates an empty queue backed by a fresh, independently collectible Myers jump-link arena.
         *
         * There is deliberately no shared global empty value: each call owns its own arena, so
         * unrelated workloads never share retained history or write contention. This factory is O(1).
         */
        public fun <T> empty(): AncestralSliceQueue<T> = empty(MyersIncrementalAncestorArena<T>())

        /**
         * Creates a Myers-backed queue holding [values] in iteration order.
         *
         * Publishes one arena node per value, for Theta(k) amortized total time on the shipped
         * backend and no ancestor query.
         *
         * @param values the values to append, captured exactly once in iteration order.
         * @throws ArithmeticException when the visible count or the ancestry depth cannot grow
         *   further.
         */
        public fun <T> from(values: Iterable<T>): AncestralSliceQueue<T> {
            var result = empty<T>()
            for (value in values) {
                result = result.addLast(value)
            }
            return result
        }
    }

    /** The number of visible values, read from the cache in O(1) without consulting the arena. */
    public val size: Int = count

    /** Whether this handle denotes an empty ancestry interval. This read is O(1). */
    public val isEmpty: Boolean
        get() = size == 0

    /**
     * The first visible value, or `null` when the queue is empty.
     *
     * The front is an ancestor selected jointly by the tail and the low boundary, so this costs one
     * ancestor query `Q(M)` rather than being an O(1) endpoint read; a singleton reuses its retained
     * tail and makes no query. A queue whose element type is nullable cannot distinguish a stored
     * `null` from an empty queue through this property; use [isEmpty] or [popFirst] for that.
     */
    public val first: T?
        get() = if (size == 0) null else arena.valueAt(firstNode())

    /**
     * The last visible value, or `null` when the queue is empty.
     *
     * This is O(1): the tail is retained directly by the handle. A queue whose element type is
     * nullable cannot distinguish a stored `null` from an empty queue through this property; use
     * [isEmpty] or [popLast] for that.
     */
    public val last: T?
        get() = if (size == 0) null else arena.valueAt(tail)

    /**
     * Returns the visible value at a zero-based [index], or `null` outside `0 until size`.
     *
     * This costs one ancestor query `Q(M)`; the last visible index reuses the retained tail and makes
     * no query. A queue whose element type is nullable cannot distinguish a stored `null` from an
     * out-of-range index through this operator; validate the index against [size] for that.
     *
     * @param index the zero-based visible index.
     * @throws ArithmeticException when the addressed absolute depth exceeds [Int.MAX_VALUE].
     */
    public operator fun get(index: Int): T? {
        if (index < 0 || index >= size) {
            return null
        }
        val node = if (index == size - 1) tail else arena.ancestorAtDepth(tail, Math.addExact(lowDepth, index))
        return arena.valueAt(node)
    }

    /**
     * Returns a queue with [value] appended below this handle's tail or empty anchor.
     *
     * This costs `U(M)` and publishes exactly one arena node. This handle and every other retained
     * version are unaffected; appending below an old handle starts a sibling branch. The count is
     * checked before the node is published, so a rejected addition publishes nothing.
     *
     * @param value the value to append.
     * @throws ArithmeticException when the visible count or the ancestry depth cannot grow further.
     */
    public fun addLast(value: T): AncestralSliceQueue<T> {
        val grown = Math.addExact(size, 1)
        return AncestralSliceQueue(arena, arena.addLeaf(tail, value), lowDepth, grown)
    }

    /**
     * Returns the queue without its first visible value, or `null` when the queue is empty.
     *
     * This is O(1) and makes no ancestor query, because only the low boundary moves. Removing the one
     * visible value of a singleton this way leaves the old tail as the empty result's anchor.
     *
     * @throws ArithmeticException when the low boundary cannot move past [Int.MAX_VALUE].
     */
    public fun removeFirst(): AncestralSliceQueue<T>? =
        if (size == 0) null else AncestralSliceQueue(arena, tail, Math.addExact(lowDepth, 1), size - 1)

    /**
     * Returns the queue without its last visible value, or `null` when the queue is empty.
     *
     * This is O(1): the new tail is the old tail's parent, which the arena reads in constant time.
     * Emptying a singleton this way anchors the result immediately before the window.
     */
    public fun removeLast(): AncestralSliceQueue<T>? =
        if (size == 0) null else AncestralSliceQueue(arena, arena.parentOf(tail), lowDepth, size - 1)

    /**
     * Removes and returns the first visible value together with the remainder, or `null` when the
     * queue is empty.
     *
     * This costs `Q(M)` because it also reports the front value. Nothing is published on either path,
     * and this handle stays valid and reusable.
     *
     * @throws ArithmeticException when the low boundary cannot move past [Int.MAX_VALUE].
     */
    public fun popFirst(): AncestralSlicePop<T>? {
        if (size == 0) {
            return null
        }
        val value = arena.valueAt(firstNode())
        return AncestralSlicePop(value, AncestralSliceQueue(arena, tail, Math.addExact(lowDepth, 1), size - 1))
    }

    /**
     * Removes and returns the last visible value together with the remainder, or `null` when the
     * queue is empty.
     *
     * This is O(1) and makes no ancestor query.
     */
    public fun popLast(): AncestralSlicePop<T>? {
        if (size == 0) {
            return null
        }
        return AncestralSlicePop(arena.valueAt(tail), AncestralSliceQueue(arena, arena.parentOf(tail), lowDepth, size - 1))
    }

    /**
     * Returns the first [count] visible values, or `null` when [count] is outside `0..size`.
     *
     * This costs `Q(M)`; `take(size)` returns this handle itself and makes no query. The result is a
     * real ancestry slice and remains appendable, and `take(0)` is anchored immediately before the
     * window.
     *
     * @param count the prefix length.
     * @throws ArithmeticException when the addressed absolute depth exceeds [Int.MAX_VALUE].
     */
    public fun take(count: Int): AncestralSliceQueue<T>? {
        if (count < 0 || count > size) {
            return null
        }
        return if (count == size) this else slice(0, count)
    }

    /**
     * Returns the queue after omitting its first [count] visible values, or `null` when [count] is
     * outside `0..size`.
     *
     * This is O(1) and makes no ancestor query, because only the low boundary moves. `drop(0)`
     * returns this handle itself, and `drop(size)` keeps the old tail as the empty result's anchor.
     *
     * @param count the number of values to omit.
     * @throws ArithmeticException when the low boundary cannot move past [Int.MAX_VALUE].
     */
    public fun drop(count: Int): AncestralSliceQueue<T>? {
        if (count < 0 || count > size) {
            return null
        }
        if (count == 0) {
            return this
        }
        return AncestralSliceQueue(arena, tail, Math.addExact(lowDepth, count), size - count)
    }

    /**
     * Returns one contiguous, appendable ancestry slice, or `null` for a range outside this queue.
     *
     * This costs `Q(M)` in general. The whole range returns this handle itself, and any range whose
     * result ends at the retained tail — every suffix — reuses that tail and makes no query. A
     * zero-length range selects the ancestor immediately before its start as an anchor, which is the
     * arena's bottom node when the start is the very front of a queue grown from an empty one.
     *
     * @param index the zero-based first visible index.
     * @param count the number of visible values in the result.
     * @throws ArithmeticException when an addressed absolute depth exceeds [Int.MAX_VALUE].
     */
    public fun slice(index: Int, count: Int): AncestralSliceQueue<T>? {
        if (index < 0 || count < 0 || count > size || index > size - count) {
            return null
        }
        if (index == 0 && count == size) {
            return this
        }

        val low = Math.addExact(lowDepth, index)
        val targetDepth = Math.addExact(low, count - 1)
        val currentTailDepth = Math.addExact(lowDepth, size - 1)
        val node = if (targetDepth == currentTailDepth) tail else arena.ancestorAtDepth(tail, targetDepth)
        return AncestralSliceQueue(arena, node, low, count)
    }

    /**
     * Splits this queue at a zero-based boundary, or returns `null` when [index] is outside `0..size`.
     *
     * This costs `Q(M)`: the prefix performs the one ancestor-seeking query and the suffix only moves
     * the low boundary. Both results are real ancestry slices and can independently start new
     * branches. A split at [size] makes no query, and neither does a split of an empty queue, whose
     * two boundaries coincide with its whole range; a split at `0` of a non-empty queue does make
     * one, because the anchored-empty rule requires the empty prefix to retain the node at depth
     * `lowDepth - 1`.
     *
     * @param index the number of visible values placed in the left result.
     * @throws ArithmeticException when an addressed absolute depth exceeds [Int.MAX_VALUE].
     */
    public fun splitAt(index: Int): AncestralSliceSplit<T>? {
        if (index < 0 || index > size) {
            return null
        }
        val left = take(index) ?: error("A validated boundary must be takeable.")
        val right = drop(index) ?: error("A validated boundary must be droppable.")
        return AncestralSliceSplit(left, right)
    }

    /**
     * Materializes the visible values in front-to-back order.
     *
     * This follows parent links from the tail into one temporary array rather than performing [size]
     * ancestor queries, taking Theta([size]) time and Theta([size]) transient element slots.
     * Concurrent appends cannot enter the captured path.
     */
    public fun toList(): List<T> {
        val values = ArrayList<T>(size)
        if (size == 0) {
            return values
        }

        var node = tail
        for (remaining in size - 1 downTo 0) {
            values.add(arena.valueAt(node))
            if (remaining != 0) {
                node = arena.parentOf(node)
            }
        }
        values.reverse()
        return values
    }

    /**
     * Returns a stable front-to-back iterator over this immutable ancestry interval.
     *
     * The iterator captures the whole path up front, so it observes exactly the handle it was
     * obtained from and never sees a later append. See [toList] for the cost.
     */
    override fun iterator(): Iterator<T> = toList().iterator()

    /** Returns whether both handles are navigated by the exact same arena object. This is O(1). */
    public fun sharesArenaWith(other: AncestralSliceQueue<T>): Boolean = arena === other.arena

    /**
     * Returns whether both handles retain the same arena and the same tail node, and therefore the
     * identical retained ancestry path. Intended for deterministic sharing tests; this is O(1).
     */
    public fun sharesTailWith(other: AncestralSliceQueue<T>): Boolean =
        arena === other.arena && tail == other.tail

    /**
     * Independently recomputes this handle's window against the arena and reports what it found.
     *
     * The validation walks [size] parent links and reads only depths, so it takes Theta([size]) time,
     * allocates O(1), and makes no ancestor query; a counting arena's query statistics are therefore
     * undisturbed.
     *
     * @throws IllegalStateException when the cached count, the low boundary, or the retained tail
     *   depth contradicts the ancestry path, including a violation of the anchored-empty rule.
     */
    public fun validateStructure(): AncestralSliceQueueStatistics {
        check(size >= 0) { "An ancestral slice queue cannot cache a negative count." }
        check(lowDepth >= 0) { "An ancestral slice window cannot begin above the arena bottom." }

        val tailDepth = arena.depthOf(tail)
        check(tailDepth == Math.addExact(lowDepth, size) - 1) {
            "The retained tail depth must equal lowDepth + count - 1."
        }

        var node = tail
        var walked = 0
        while (walked < size) {
            check(arena.depthOf(node) == tailDepth - walked) {
                "An ancestry path depth must decrease by exactly one per parent link."
            }
            walked++
            if (walked < size) {
                node = arena.parentOf(node)
            }
        }
        if (size != 0) {
            check(arena.depthOf(node) == lowDepth) {
                "The oldest visible node must sit at the window's low depth."
            }
        }

        return AncestralSliceQueueStatistics(size, lowDepth, tailDepth, walked, arena.publishedNodeCount)
    }

    /**
     * Returns the constant-sized window header, mirroring the reference debugger display.
     *
     * This is O(1) and reads no values; use [toList] to render the visible contents.
     */
    override fun toString(): String = "AncestralSliceQueue(size=$size, lowDepth=$lowDepth, tail=$tail)"

    private fun firstNode(): Int = if (size == 1) tail else arena.ancestorAtDepth(tail, lowDepth)
}
